File: examples/python/migration_guides/mergent.py
from datetime import datetime, timedelta, timezone
from typing import Any, Dict, List, Mapping

import requests
from pydantic import BaseModel
from requests import Response

from hatchet_sdk.context.context import Context

from .hatchet_client import hatchet


async def process_image(image_url: str, filters: List[str]) -> Dict[str, Any]:
    # Do some image processing
    return {"url": image_url, "size": 100, "format": "png"}


# > Before (Mergent)
async def process_image_task(request: Any) -> Dict[str, Any]:
    image_url = request.json["image_url"]
    filters = request.json["filters"]
    try:
        result = await process_image(image_url, filters)
        return {"success": True, "processed_url": result["url"]}
    except Exception as e:
        print(f"Image processing failed: {e}")
        raise




# > After (Hatchet)
class ImageProcessInput(BaseModel):
    image_url: str
    filters: List[str]


class ImageProcessOutput(BaseModel):
    processed_url: str
    metadata: Dict[str, Any]


@hatchet.task(
    name="image-processor",
    retries=3,
    execution_timeout="10m",
    input_validator=ImageProcessInput,
)
async def image_processor(input: ImageProcessInput, ctx: Context) -> ImageProcessOutput:
    # Do some image processing
    result = await process_image(input.image_url, input.filters)

    if not result["url"]:
        raise ValueError("Processing failed to generate URL")

    return ImageProcessOutput(
        processed_url=result["url"],
        metadata={
            "size": result["size"],
            "format": result["format"],
            "applied_filters": input.filters,
        },
    )




async def run() -> None:
    # > Running a task (Mergent)
    headers: Mapping[str, str] = {
        "Authorization": "Bearer <token>",
        "Content-Type": "application/json",
    }

    task_data = {
        "name": "4cf95241-fa19-47ef-8a67-71e483747649",
        "queue": "default",
        "request": {
            "url": "https://example.com",
            "headers": {
                "Authorization": "fake-secret-token",
                "Content-Type": "application/json",
            },
            "body": "Hello, world!",
        },
    }

    try:
        response: Response = requests.post(
            "https://api.mergent.co/v2/tasks",
            headers=headers,
            json=task_data,
        )
        print(response.json())
    except Exception as e:
        print(f"Error: {e}")

    # > Running a task (Hatchet)
    result = await image_processor.aio_run(
        ImageProcessInput(image_url="https://example.com/image.png", filters=["blur"])
    )

    # you can await fully typed results
    print(result)


async def schedule() -> None:
    # > Scheduling tasks (Mergent)
    options = {
        # same options as before
        "json": {
            # same body as before
            "delay": "5m"
        }
    }

    print(options)

    # > Scheduling tasks (Hatchet)
    # Schedule the task to run at a specific time
    run_at = datetime.now(tz=timezone.utc) + timedelta(days=1)
    await image_processor.aio_schedule(
        run_at,
        ImageProcessInput(image_url="https://example.com/image.png", filters=["blur"]),
    )

    # Schedule the task to run every hour
    await image_processor.aio_create_cron(
        "run-hourly",
        "0 * * * *",
        ImageProcessInput(image_url="https://example.com/image.png", filters=["blur"]),
    )

File: examples/go/migration-guides/mergent.go
package mergent

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	hatchet "github.com/hatchet-dev/hatchet/sdks/go"
	"github.com/hatchet-dev/hatchet/sdks/go/features"
)

// ProcessImage simulates image processing
func ProcessImage(imageURL string, filters []string) (map[string]interface{}, error) {
	// Do some image processing
	return map[string]interface{}{
		"url":    imageURL,
		"size":   100,
		"format": "png",
	}, nil
}

// > Before (Mergent)
type MergentRequest struct {
	ImageURL string   `json:"image_url"`
	Filters  []string `json:"filters"`
}

type MergentResponse struct {
	Success      bool   `json:"success"`
	ProcessedURL string `json:"processed_url"`
}

func ProcessImageMergent(req MergentRequest) (*MergentResponse, error) {
	result, err := ProcessImage(req.ImageURL, req.Filters)
	if err != nil {
		return nil, err
	}

	return &MergentResponse{
		Success:      true,
		ProcessedURL: result["url"].(string),
	}, nil
}


// > After (Hatchet)
type ImageProcessInput struct {
	ImageURL string   `json:"image_url"`
	Filters  []string `json:"filters"`
}

type ImageProcessOutput struct {
	ProcessedURL string `json:"processed_url"`
	Metadata     struct {
		Size           int      `json:"size"`
		Format         string   `json:"format"`
		AppliedFilters []string `json:"applied_filters"`
	} `json:"metadata"`
}

func ImageProcessor(client *hatchet.Client) *hatchet.StandaloneTask {
	processor := client.NewStandaloneTask("image-processor",
		func(ctx hatchet.Context, input ImageProcessInput) (*ImageProcessOutput, error) {
			result, err := ProcessImage(input.ImageURL, input.Filters)
			if err != nil {
				return nil, fmt.Errorf("processing image: %w", err)
			}

			if result["url"] == "" {
				return nil, fmt.Errorf("processing failed to generate URL")
			}

			output := &ImageProcessOutput{
				ProcessedURL: result["url"].(string),
				Metadata: struct {
					Size           int      `json:"size"`
					Format         string   `json:"format"`
					AppliedFilters []string `json:"applied_filters"`
				}{
					Size:           result["size"].(int),
					Format:         result["format"].(string),
					AppliedFilters: input.Filters,
				},
			}

			return output, nil
		},
	)

	// Example of running a task
	_ = func() error {
		// > Running a task
		result, err := processor.Run(context.Background(), ImageProcessInput{
			ImageURL: "https://example.com/image.png",
			Filters:  []string{"blur"},
		})
		if err != nil {
			return err
		}

		fmt.Printf("Result: %+v\n", result)

		return nil
	}

	// Example of registering a task on a worker
	_ = func() error {
		// > Declaring a Worker
		w, err := client.NewWorker("image-processor-worker", hatchet.WithWorkflows(processor))
		if err != nil {
			return err
		}

		err = w.StartBlocking(context.Background())
		if err != nil {
			return err
		}

		return nil
	}

	return processor
}

func RunMergentTask() error {
	return nil
}

func RunningTasks(client *hatchet.Client) error {
	// > Running a task (Mergent)
	task := struct {
		Request struct {
			URL     string            `json:"url"`
			Body    string            `json:"body"`
			Headers map[string]string `json:"headers"`
		} `json:"request"`
		Name  string `json:"name"`
		Queue string `json:"queue"`
	}{
		Request: struct {
			URL     string            `json:"url"`
			Body    string            `json:"body"`
			Headers map[string]string `json:"headers"`
		}{
			URL: "https://example.com",
			Headers: map[string]string{
				"Authorization": "fake-secret-token",
				"Content-Type":  "application/json",
			},
			Body: "Hello, world!",
		},
		Name:  "4cf95241-fa19-47ef-8a67-71e483747649",
		Queue: "default",
	}

	taskJSON, err := json.Marshal(task)
	if err != nil {
		return fmt.Errorf("marshaling task: %w", err)
	}

	req, err := http.NewRequest(http.MethodPost, "https://api.mergent.co/v2/tasks", bytes.NewBuffer(taskJSON))
	if err != nil {
		return fmt.Errorf("creating request: %w", err)
	}

	req.Header.Add("Authorization", "Bearer <API_KEY>")
	req.Header.Add("Content-Type", "application/json")

	httpClient := &http.Client{}
	res, err := httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("sending request: %w", err)
	}
	defer res.Body.Close()

	fmt.Printf("Mergent task created with status: %d\n", res.StatusCode)

	// > Running a task (Hatchet)
	processor := ImageProcessor(client)

	result, err := processor.Run(context.Background(), ImageProcessInput{
		ImageURL: "https://example.com/image.png",
		Filters:  []string{"blur"},
	})
	if err != nil {
		return err
	}
	fmt.Printf("Result: %+v\n", result)

	// > Scheduling tasks (Hatchet)
	// Schedule the task to run at a specific time
	scheduleRef, err := client.Schedules().Create(
		context.Background(),
		"image-processor",
		features.CreateScheduledRunTrigger{
			TriggerAt: time.Now().Add(time.Second * 10),
			Input: map[string]interface{}{
				"image_url": "https://example.com/image.png",
				"filters":   []string{"blur"},
			},
		},
	)
	if err != nil {
		return err
	}

	// or schedule to run every hour
	cronRef, err := client.Crons().Create(
		context.Background(),
		"image-processor",
		features.CreateCronTrigger{
			Name:       "run-hourly",
			Expression: "0 * * * *",
			Input: map[string]interface{}{
				"image_url": "https://example.com/image.png",
				"filters":   []string{"blur"},
			},
		},
	)
	if err != nil {
		return err
	}

	fmt.Printf("Scheduled tasks with refs: %+v, %+v\n", scheduleRef, cronRef)
	return nil
}
